### PR TITLE
Code adapt.v1.1

### PR DESCRIPTION
**Pull request by galtimur**

### CODE REVIEW

File: retro_pytorch/train_functions.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    print("------ Validation ------")
+    losses_val_cur = []
+    for seq, docs in tqdm(aggregate, ncols=80):
+        loss = calc_loss(seq, docs, model, no_retrieve, fetch_neighbours)
+        losses_val_cur.append(loss.item())
+
+    return losses_val_cur
+
+
+def val_upadate(
```
Comment (egor-bogomolov) [inline retro_pytorch/train_functions.py]: rename to `val_update`

File: scripts/train.py [hunk 2]
```diff
@@ -1,40 +1,54 @@
...
+model_name = paths.model_name + add_flag
+train_data_path = paths.data_folder + paths.train_data_file
+val_data_path = paths.data_folder + paths.val_data_file
+filename_train = paths.out_folder + paths.out_filename_train + add_flag + ".txt"
+filename_val = paths.out_folder + paths.out_filename_val + add_flag + ".txt"
+f_train = open(filename_train, "a")
+f_val = open(filename_val, "a")
 
+## TODO I am not sure that at this step I want to move model hyperparameters to a config file - it is easier for me to read it here.
+# instantiate RETRO, fit it into the TrainingWrapper with correct settings
```
Comment (egor-bogomolov) [inline scripts/train.py]: Yet you cannot try different configurations without rewriting the file or duplicating code. Config will look exactly like this, and here you would do something like `retro = RETRO(**config.model_parameters)`

File: scripts/train.py [hunk 3]
```diff
@@ -53,162 +67,113 @@
...
     retro=retro,  # path to retro instance
     knn=2,  # knn (2 in paper was sufficient)
     chunk_size=64,  # chunk size (64 in paper)
-    documents_path=data_folder,  # path to folder of text
+    documents_path=paths.data_folder,  # path to folder of text
     data_file_paths=[],
-    chunks_memmap_path=texts_folder + "train.chunks.dat",  # path to chunks
-    seqs_memmap_path=texts_folder + "train.seq.dat",  # path to sequence data
-    doc_ids_memmap_path=texts_folder
+    chunks_memmap_path=paths.texts_folder + "train.chunks.dat",  # path to chunks
```
Comment (egor-bogomolov) [inline scripts/train.py]: Consider using os.path.join to join paths, it will save you from accidentally getting paths like "path/to/foldertrain.chunks.dat"

File: scripts/train.py [hunk 4]
```diff
@@ -53,162 +67,113 @@
...
     knn_extra_neighbors=100,  # num extra neighbors to fetch
+    precalculate_knn=False,
     max_index_memory_usage="10G",
     current_memory_available="32G",
 )
 
 # %%
 
+### setting up number of steps. freq_val - frequency of validation, num_val - number of validation steps
+freq_val = 1000
```
Comment (egor-bogomolov) [inline scripts/train.py]: It would be nice to have these numbers in the configuration file

File: retro_pytorch/train_functions.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    return batch_aggregate, val_step
+
+
+def val_steps(
+    model: Any,
+    no_retrieve: bool,
+    fetch_neighbours: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    aggregate: list[tuple[torch.Tensor, torch.Tensor]],
+) -> list[float]:
+    model.eval()
```
Comment (egor-bogomolov) [inline retro_pytorch/train_functions.py:62]: You should also use `torch.no_grad()` or `torch.inference_mode()` to avoid gradient computations. Also, you do backprop from loss even if you call calc_loss in validation, please avoid it.

File: scripts/train.py [hunk 6]
```diff
@@ -53,162 +67,113 @@
...
+train_ds = DatasetJsonl(train_data_path, cnunk_size=64, seq_length=512, pad_id=0)
+val_ds = DatasetJsonl(val_data_path, cnunk_size=64, seq_length=512, pad_id=0)
+train_dl = DataLoaderFromFile(train_ds, batch_size=batch_size)
+val_dl = DataLoaderFromFile(val_ds, batch_size=batch_size)
 
 optim, scheduler = wrapper_db.get_optimizer(warmup_steps=warmup_steps, training_steps=total_steps, lr=lr, wd=0.01)
 scheduler.step()
 fetch_neighbours = wrapper_db.fetch_neighbours
+fetch_random_chunk = wrapper_db.fetch_random_chunk
+generate_pure_random_chunk = wrapper_db.fetch_random_chunk
```
Comment (egor-bogomolov) [inline scripts/train.py]: I guess these two should be different

**Review by egor-bogomolov — CHANGES_REQUESTED** [review]
Left a few comments, please fix training on the validation data and usage of random retrieval functions 🙂
